source/extensions/http/cache/file_system_http_cache/cache_eviction_thread.cc

```diff
@@ -0,0 +1,77 @@
+#include "source/extensions/http/cache/file_system_http_cache/cache_eviction_thread.h"
+
+#include "envoy/thread/thread.h"
+
+#include "source/extensions/http/cache/file_system_http_cache/file_system_http_cache.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+namespace FileSystemHttpCache {
+
+CacheEvictionThread::CacheEvictionThread(Thread::ThreadFactory& thread_factory)
+    : thread_factory_(thread_factory) {}
+
+void CacheEvictionThread::addCache(FileSystemHttpCache& cache) {
+  absl::MutexLock lock(&cache_mu_);
+  if (caches_.empty()) {
+    thread_ = thread_factory_.createThread([this]() { work(); });
+  }
+  bool inserted = caches_.emplace(&cache).second;
+  ASSERT(inserted);
+}
+
+void CacheEvictionThread::removeCache(FileSystemHttpCache& cache) {
+  absl::MutexLock lock(&cache_mu_);
+  bool removed = caches_.erase(&cache);
+  ASSERT(removed);
+  if (caches_.empty()) {
+    terminate();
+    thread_->join();
+  }
+}
+
+void CacheEvictionThread::signal() {
+  absl::MutexLock lock(&mu_);
+  signalled_ = true;
+}
+
+void CacheEvictionThread::terminate() {
+  absl::MutexLock lock(&mu_);
+  terminating_ = true;
+  signalled_ = true;
+}
+
+bool CacheEvictionThread::waitForSignal() {
+  absl::MutexLock lock(&mu_);
+  // Worth noting here that if `signalled_` is already true, the lock is not released
+  // until idle_ is false again, so waitForIdle will not return until `signalled_`
+  // stays false for the duration of an eviction cycle.
+  idle_ = true;
+  mu_.Await(absl::Condition(&signalled_));
+  signalled_ = false;
+  idle_ = false;
+  return !terminating_;
+}
+
+void CacheEvictionThread::work() {
+  while (waitForSignal()) {
+    absl::MutexLock lock(&cache_mu_);
```

Comment:
this lock might be held for a long time while the cache is iterated over. I'm not sure if this is a real problem.

One suggestion would be to have a locked initial loop that makes a local vector of shared_ptr<FileCache> to hold onto them. Then you can release the lock and iterate over the caches.

Reply:
Added a comment about this. Releasing the lock first is no good because when you're not holding the lock, the pointers can become invalid.

Reply:
The thread can't own the caches with shared_ptr because the caches indirectly own the thread with shared_ptr. It would make a memory leak loop.

The caches are solely owned by filter config, which is necessary as that's the only way to manage removal of unused caches when the config changes.

weak_ptr transitioning to shared_ptr seems like an option, but if you do *that* then you can end up with the CacheEvictionThread being the owner of the last reference to the last cache, which means when that block goes out of scope the CacheEvictionThread destructor is called, which means it tries to call join from the thread itself, and *that* was the problem that led to the non-constructor-based initialization before, which didn't actually help because removeCache is still part of the cache destructor.

Reply:
I can see how this has gotten complex :)

The high level requirement (from my perspective) is that I think we shouldn't get into a situation where we are blocking the data plane on a cache cleanup.

You might be able to use atomics to help you avoid that, but it requires some careful thinking.

Reply:
Done by splitting off a shared *part* of the cache instance, which allows the main cache instance to still be deleted by its owner, but the eviction thread to keep alive the part that it's using until it's done with it.